src/js/components/Carousel/Carousel.js

```diff
@@ -1,242 +1,218 @@
+      if (inTransition) {
+        const { offsetWidth: currentWidth, offsetHeight: currentHeight } =
+          containerRef.current;
+        if (currentHeight && currentWidth) {
+          setContainerProps({
+            containerHeight: currentHeight,
+            containerWidth: currentWidth,
           });
-          onChildChange(0);
         }
-      }, play);
-
-      timerRef.current = timer;
-
-      return () => {
-        clearTimeout(timer);
-      };
+      } else {
+        setContainerProps({
+          containerHeight: undefined,
+          containerWidth: undefined,
+        });
+      }
     }
-    return () => {};
-  }, [activeIndex, play, children, lastIndex, onChildChange]);
+  }, [inTransition, noDimensions]);
 
-  const onRight = () => {
-    if (activeIndex >= lastIndex) {
-      return;
-    }
-    clearInterval(timerRef.current);
-    setIndexes({
-      activeIndex: activeIndex + 1,
-      priorActiveIndex: activeIndex,
-    });
-    onChildChange(activeIndex + 1);
-  };
-
-  const onLeft = () => {
-    if (activeIndex <= 0) {
-      return;
-    }
-    clearInterval(timerRef.current);
-    setIndexes({
-      activeIndex: activeIndex - 1,
-      priorActiveIndex: activeIndex,
-    });
-    onChildChange(activeIndex - 1);
-  };
-
-  const onSelect = (index) => () => {
-    if (activeIndex !== index) {
-      clearInterval(timerRef.current);
-      setIndexes({ activeIndex: index, priorActiveIndex: activeIndex });
-      onChildChange(index);
+  /**
+   * Delays the transitions between Carousel slides. This is needed to
+   * avoid users "spamming" the controls which results in jarring animations
+   * and a bad user experience.
+   */
+  useEffect(() => {
+    let transitionTimer;
+    if (inTransition) {
+      transitionTimer = setTimeout(() => {
+        setInTransition(false);
+      }, animationDuration);
     }
-  };
+    return () => clearTimeout(transitionTimer);
+  }, [inTransition, setInTransition, animationDuration]);
 
-  const showArrows = controls && controls !== 'selectors';
-  const showSelectors = controls && controls !== 'arrows';
-
-  const CurrentIcon = theme.carousel.icons.current;
-  const iconColor = normalizeColor(
-    theme.carousel.icons.color || 'control',
-    theme,
-  );
-
-  const selectors = [];
-  const wrappedChildren = Children.map(children, (child, index) => {
-    selectors.push(
-      <Button
-        a11yTitle={`Show carousel slide ${index + 1}`}
-        // eslint-disable-next-line react/no-array-index-key
-        key={index}
-        icon={
-          <CurrentIcon color={activeIndex === index ? iconColor : undefined} />
-        }
-        onClick={onSelect(index)}
-      />,
-    );
+  // Handles auto-playing Carousel slides
+  useEffect(() => {
+    let playTimer;
+    if (play) {
+      playTimer = setInterval(() => {
+        onNext();
+      }, play);
+    }
+    return () => clearTimeout(playTimer);
+  }, [current, inTransition, play, onNext]);
```

Comment:
don't need current or inTransition here although you're basically getting that since onNext's useCallback is dependent on both. The result is this useEffect destroys and re-creates the timer each transition.